### PR TITLE
feat: surfaces experiment

### PR DESCRIPTION
**Pull request by zhenyasav**
### Summary
📝🆕🔧

This pull request refactors some types, adds some documentation, and introduces a new way of handling actions in the `Surface` framework. It improves the readability, consistency, and expressiveness of the code and the design of the framework. It also documents some of the developer feedback and questions that are relevant for the framework development.

> _We're working on the Surface framework, me hearties_
> _We're adding notes and types and actions, aye_
> _We're refactoring `T` to `TState`, me lads and lasses_
> _We're heaving on the code on the count of three, hooray_

### Walkthrough
*  Introduce a new way of defining and handling actions in the Surface framework using strongly typed actions and handlers ([link](https://github.com/dxos/dxos/pull/3148/files?diff=unified&w=0#diff-99d67a75f05fbf7ecc5922d2db58305b5bcd561040813b8b09f32834147a8102R1-R42))
* Rename the generic type parameter `T` to `TState` in the `App.tsx` file to make it more descriptive and consistent ([link](https://github.com/dxos/dxos/pull/3148/files?diff=unified&w=0#diff-29a0b81d5dc1cc789146e7854827a89a3f2a9f1d6706b38e741a96f04dfbb067L17-R38), [link](https://github.com/dxos/dxos/pull/3148/files?diff=unified&w=0#diff-29a0b81d5dc1cc789146e7854827a89a3f2a9f1d6706b38e741a96f04dfbb067L44-R45))
* Add some notes to the `README.md` file from Zhenya about the design choices and questions of the Surface framework ([link](https://github.com/dxos/dxos/pull/3148/files?diff=unified&w=0#diff-48c510d98912c40a55e794ac2389ed6c2b66f1e4356148bb5c624f8bd20c903aR2-R8))

### CODE REVIEW

File: packages/experimental/surface/src/framework/PluginContext.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  plugins: Plugin[];
+};
+
+const defaultContext: PluginContextValue = { plugins: [] };
+
+const PluginContext = createContext<PluginContextValue>(defaultContext);
+
+export const usePluginContext = () => useContext(PluginContext);
+
+export const PluginContextProvider = ({ plugins: definitions }: { plugins: PluginDefinition[] }) => {
```
Comment (thure) [inline packages/experimental/surface/src/framework/PluginContext.tsx:20]: Just `PluginProvider` would make the most sense.

File: packages/experimental/surface/src/plugins/ThemePlugin.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import { ThemeProvider } from '@dxos/aurora';
+
+import { definePlugin } from '../framework';
+
+import '@dxosTheme';
+
+export const ThemePlugin = definePlugin({
+  meta: {
+    id: 'dxos:ThemePlugin'
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/ThemePlugin.tsx:15]: perhaps `org.dxos/plugin/ThemePlugin` we need a GUID for registration

File: packages/experimental/surface/src/plugins/RoutesPlugin.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+export type RouterPluginProvides = {
+  router: {
+    routes: () => RouteObject[];
+  };
+};
+
+export type RouterPlugin = Plugin<RouterPluginProvides>;
+
+const routerPlugins = (plugins: Plugin[]): RouterPlugin[] => {
+  return (plugins as RouterPlugin[]).filter((p) => p.provides?.router);
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/RoutesPlugin.tsx:25]: this feels unsafe and/or cumbersome? can we make the interface concrete/introspectable?
for inspiration: https://learn.microsoft.com/en-us/cpp/atl/queryinterface?view=msvc-170

File: packages/experimental/surface/src/plugins/RoutesPlugin.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+export const RoutesPlugin = definePlugin({
+  meta: {
+    id: 'RoutesPlugin'
+  },
+  provides: {
+    context: ({ children }) => <HashRouter>{children}</HashRouter>,
+    components: {
+      default: () => {
+        const { plugins } = usePluginContext();
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/RoutesPlugin.tsx:36]: I'd like to see a fuller example re how separate plugins can cooperate to build a set of consistent routes for a given app. my intuition is that kind of coupling belongs in the app itself.

File: packages/experimental/surface/src/plugins/SpacePlugin.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  },
+  provides: {
+    router: {
+      routes: () => [
+        {
+          path: '/space/:spaceId',
+          element: (
+            <Surface
+              component='dxos:SplitViewPlugin/SplitView'
+              surfaces={{
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/SpacePlugin.tsx:35]: the nesting seems odd.
shouldn't the embedded component decide about the layout of internal surfaces and rely on plugins to populate that.

File: packages/experimental/surface/src/plugins/ListViewPlugin.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  id: string;
+  label: string;
+  icon?: FC;
+  invoke: (event: UIEvent) => MaybePromise<void>;
+};
+
+export type GraphPluginProvides = {
+  graph: {
+    nodes?: (plugins: Plugin[], parent?: GraphNode) => MulticastObservable<GraphNode[]>;
+    actions?: (plugins: Plugin[], parent?: GraphNode) => MulticastObservable<GraphNodeAction[]>;
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/ListViewPlugin.tsx:61]: actions are labelled event sinks? i.e., "things that can be done to a particular node?"

File: packages/experimental/surface/src/plugins/ListViewPlugin.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  actions?: GraphNodeAction[];
+  children?: GraphNode[];
+  parent?: GraphNode;
+};
+
+type GraphNodeAction = {
+  id: string;
+  label: string;
+  icon?: FC;
+  invoke: (event: UIEvent) => MaybePromise<void>;
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/ListViewPlugin.tsx:55]: it feel that the event payload should be typed and not mixed with React events. (and perhaps have a `source` property this it the incipient low-level event).

File: packages/experimental/surface/src/plugins/SpacePlugin.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      SpaceContainer
+    },
+    graph: {
+      nodes: (plugins, parent) => {
+        const clientPlugin = findPlugin<ClientPluginProvides>(plugins, 'dxos:ClientPlugin');
+        if (!clientPlugin) {
+          return MulticastObservable.of([]);
+        }
+
+        if (parent) {
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/SpacePlugin.tsx:54]: comment needed

File: packages/experimental/surface/src/plugins/SpacePlugin.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          return MulticastObservable.of([]);
+        }
+
+        return clientPlugin.provides.client.spaces.map((spaces) =>
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/SpacePlugin.tsx:58]: doesn't this respond with a static (non reactive) array?

File: packages/experimental/surface/src/plugins/ClientPlugin.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    id: 'dxos:ClientPlugin'
+  },
+  init: async () => {
+    const client = new Client();
+    await client.initialize();
+
+    return {
+      client,
+      context: ({ children }) => <ClientProvider client={client}>{children}</ClientProvider>,
+      graph: {
```
Comment (richburdon) [inline packages/experimental/surface/src/plugins/ClientPlugin.tsx:28]: this feels like a cyclic dep. client is really far down our stack, but this is "providing" very high level (opinionated) graph plugin semantics?

File: packages/experimental/surface/src/framework/Context.tsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import React, { FC, PropsWithChildren } from 'react';
+
+import { Plugin } from './Plugin';
+
+export const compose = (contexts: FC<PropsWithChildren>[]) => {
```
Comment (richburdon) [inline packages/experimental/surface/src/framework/Context.tsx:9]: I know we've discussed -- you've explored -- both tracks here so I'm curious how you landed back with composable contexts?

we have a root plugin context. what additional benefit do we get from each plugin providing a unique context. 
my understanding of contexts is that they just manage a value defined by the provider. what's the advantage here over managing that value (by plugin type) in a root plugin provider?

File: packages/experimental/surface/src/plugins/ListViewPlugin.tsx [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+export type GraphPluginProvides = {
+  graph: {
+    nodes?: (plugins: Plugin[], parent?: GraphNode) => MulticastObservable<GraphNode[]>;
```
Comment (dmaretskyi) [inline packages/experimental/surface/src/plugins/ListViewPlugin.tsx:60]: is `plugins` the list of other registered plugins?

maybe we should extract it to some common `env` or `context` object

File: packages/experimental/surface/src/framework/Context.tsx (hunk 11 above)
Comment (dmaretskyi) [reply packages/experimental/surface/src/framework/Context.tsx:9]: my intuition here is that storing plugin state in a class backed by reactivity that @wittjosiah is doing right now would be simpler

**Comment by wittjosiah** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3148** <a href="https://app.graphite.dev/github/pr/dxos/dxos/3148" target="_blank">Graphite</a>  👈
    * **PR #3228** <a href="https://app.graphite.dev/github/pr/dxos/dxos/3228" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dxos/dxos/3148?utm_source=stack-comment).

**Pull request merged**